sklearn/datasets/_svmlight_format_fast.pyx

```diff
@@ -113,3 +116,128 @@ def _load_svmlight_file(f, dtype, bint multilabel, bint zero_based,
             break
 
     return (dtype, data, indices, indptr, labels, query)
+
+ctypedef fused int_or_float1:
+    cython.integral
+    cython.floating
+    signed long long
+
+ctypedef fused int_or_float2:
+    signed long long
+    double
+
+ctypedef fused int_or_longlong:
+    cython.integral
+    signed long long
+
+def get_dense_row_string(int_or_float1[:,:] X, Py_ssize_t[:] x_inds, int_or_float2[:] x_vals, Py_ssize_t row, str value_pattern, bint one_based):
+    cdef:
+        Py_ssize_t row_length = X.shape[1]
+        Py_ssize_t x_nz_used = 0
+        Py_ssize_t k
+
+    for k in range(row_length):
+        val = X[row,k]
+        if val!=0:
+            x_inds[x_nz_used] = k
+            x_vals[x_nz_used] = val
+            x_nz_used += 1
+    return " ".join(value_pattern % (j+one_based, val) for i, (j, val) in enumerate(zip(x_inds, x_vals)) if i < x_nz_used)
+
+def get_sparse_row_string(int_or_float1[:] X_data, int[:] X_indptr, int[:] X_indices, Py_ssize_t row, str value_pattern, bint one_based):
+    cdef:
+        Py_ssize_t row_start = X_indptr[row]
+        Py_ssize_t row_end = X_indptr[row+1]
+
+    return " ".join(value_pattern % (X_indices[i]+one_based, X_data[i]) for i in range(row_start, row_end))
+
+def _dump_svmlight_file(X, y, f, bint multilabel, bint one_based, int_or_longlong[:] query_id, bint X_is_sp, bint y_is_sp):
+    cdef bint X_is_integral
+    if X_is_sp:
+        X_is_integral = X.data.dtype.kind == "i"
+    else:
+        X_is_integral = X.dtype.kind == "i"
+    if X_is_integral:
+        value_pattern = "%d:%d"
+    else:
+        value_pattern = "%d:%.16g"
+    if y.dtype.kind == "i":
+        label_pattern = "%d"
+    else:
+        label_pattern = "%.16g"
+
+    line_pattern = "%s"
+    if query_id is not None:
+        line_pattern += " qid:%d"
+    line_pattern += " %s\n"
+
+    cdef:
+        Py_ssize_t num_labels = y.shape[1]
+        Py_ssize_t x_len = X.shape[0]
+        Py_ssize_t row_length = X.shape[1]
+        Py_ssize_t i
+        Py_ssize_t j
+        Py_ssize_t col_start
+        Py_ssize_t col_end
+        bint first
+        array.array[signed long long] int_template = array.array('q',[])
+        array.array[double] float_template = array.array('d',[])
+        Py_ssize_t x_nz_used
+        cnp.ndarray _x_inds = np.empty(row_length, dtype=np.intp)
+        Py_ssize_t[:] x_inds = _x_inds
+        array.array x_vals
+        signed long long[:] x_vals_int
+        double[:] x_vals_float
+
+    if not X_is_sp:
+        if X_is_integral:
+            x_vals = array.clone(int_template, row_length, False)
```

Comment:
Since `x_vals` is a fixed size, I do not think we need to use `array.array` anymore. Can it be an `ndarray` of size `row_length`?

```python
        if X_is_integral:
            x_vals_int = np.zeros(row_length, dtype=np.longlong)
        else:
            x_vals_float = np.zeros(row_length, dtype=np.float64)
```

And then in the loop:

```python
            if X_is_integral:
                s = get_dense_row_string(X, x_inds, x_vals_int, i, value_pattern, one_based)
            else:
                s = get_dense_row_string(X, x_inds, x_vals_float, i, value_pattern, one_based)
```

Reply:
Good idea -- done!